File: problem215.cs
public class Solution {
    public int FindKthLargest(int[] nums, int k) {
        IList<int> smaller = new List<int>();
        IList<int> middle = new List<int>();
        IList<int> larger = new List<int>();
        Random rand = new Random();
        int pivot = rand.Next(0, nums.Length);
        var start = nums[pivot];
        for (var i = 0; i < nums.Length; i++) {
            if (nums[i] == start) {
                middle.Add(nums[i]);
            } else if (nums[i] > start) {
                larger.Add(nums[i]);
            } else {
                smaller.Add(nums[i]);
            }
        }

        if (k <= larger.Count()) {
            return FindKthLargest(larger.ToArray(), k);
        } else if (k - larger.Count() <= middle.Count()) {
            return start;
        } else {
            return FindKthLargest(smaller.ToArray(), k - larger.Count() - middle.Count());
        }
    }
}

File: su19/problem215.py
class Solution(object):
    def findKthLargest(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        import random
        pivot = nums[random.randint(0, len(nums)-1)]
        less = []
        eq = [] 
        greater = []
        for num in nums:
            if num < pivot:
                less.append(num)
            elif num == pivot:
                eq.append(num)
            else:
                greater.append(num)
        if k <= len(greater):
            return self.findKthLargest(greater, k)
        elif k > len(greater) + len(eq):
            return self.findKthLargest(less, k - len(greater) - len(eq))
        else:
            return pivot